README.md
# <img src="caravel/static/caravel.svg" alt="caravel logo" height="70" align="left"/> caravel

[![PEP compatible](http://pepkit.github.io/img/PEP-compatible-green.svg)](http://pepkit.github.io)

`caravel` is a local web GUI for [looper](https://looper.readthedocs.io/). Caravel gives you a slick web interface to interact with your projects formatted as [PEPs](http://pepkit.github.io). You can use it to submit jobs to any cluster resource manager, monitor jobs, summarize results, and browse project summary web pages. `caravel` basically builds on top of `looper`, but uses the lightweight `flask` microframework to give you a web interface, making it even easier to manage your jobs for those who like pointy-clicky.

Documentation is hosted at [caravel.databio.org](http://caravel.databio.org) (source in the [/docs](/docs) folder).

## docker

[![Build Status](https://travis-ci.org/pepkit/caravel.svg?branch=master)](https://travis-ci.org/pepkit/caravel)

A docker image for running caravel is available through the dockerfile in this repository. To build the image:

```
docker build -t caravel_demo .
```

Run the container with `caravel --demo`:

```
docker run -p 5000:5000 caravel_demo caravel --demo
```


docs/changelog.md
# Changelog

This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html) and [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) format.

## [0.13.3] -- unreleased

### Added:

- possibility to execute library module as a script: `python -m caravel ...`

## [0.13.2] -- 2019-12-13

### Changed: 

- caravel config file format change. introduced in v0.2:
    - `config_version` entry
    - a possibility to add project-level attributes (like `name`) 
- removed custom command feature
- list of projects in the index page to a table representation
- removed the subproject switching functionality on the project page
  
### Added:

- `CaravelConf` class for caravel config file management
- the possibility to set preferences in the config file
- direct subproject activation from the index page
- text indicating currently active subproject


## [0.13.1] -- 2019-07-18

### Fixed:

 - problem creating Project metadata regarding defined subprojects when none are defined; [#120](https://github.com/pepkit/caravel/issues/120)
 - problem with navbar dropdowns not expanding; [#121](https://github.com/pepkit/caravel/issues/121)
 - error when summarizing a project that has not been run; [#111](https://github.com/pepkit/caravel/issues/111)

### Changed: 

 - computing configuration page to preferences page
 - status page produced by `looper summarize` is not included in the navbar anymore
 - Project metadata is now hidden in a modal
 - the files to be deleted are not previewed beforehand
 - switched to list representation of the available projects in the "Home" page
  
### Added:

 - demonstrational data
 - demo mode when run with `--demo` option
 - go to the top buttons where relevant
 - "live" monitoring of the submitted jobs in the sample status table
 - moved sample status to the "Process" page
 - "other settings" tab in the preferences page

## [0.13.0] -- 2019-05-02

### Fixed:

  - silent failing of `looper rerun`
  - `looper summarize` not overwriting the existing summary-related files

### Changed: 

  - `lump` and `lumpn` defaults to 0 and 1, respectively
  
### Added:
  
  - custom summarizers support
  - subproject activation and deactivation updates all the `Project` object attributes in the "Process" page
  - link to the computing configuration settings in the "Home" page

## [0.2.0] -- 2019-04-18

### Changed:

  - Use logger instead of printing to the error for logging
  - Instead of having the optional `token` section in the config file 
  allow to use hidden `.caravel_token` to predefine the token
  - Change logging levels to 10 (debug) and 20 (info) in debug and normal modes, respectively

### Added:

  - New `bootstrap` UI
  - Back button to each page
  - Compute environment configuration page
  - Deactivate subproject with a dedicated button
  - Use `divvy` for computing environment configuration
  - Display `looper` log in the terminal and in a separate page, when finished
  - `looper` and `peppy` versions assurance
  - New way of `caravel`, `looper` and `python` versions monitoring on the client side in the footer
  - Systematic HTML form elements determination based on `looper` arguments options
  - The port that the server will be run on can be determined manually from the CLI (`-p` and `--port` options)
  - Seamless integration with `looper` reporting module, the pages are displayed with `caravel`'s navbar and footer

## [0.1.1] -- 2019-01-15

### Changed:

  - Fixed a bug that prevented command options from populating (caravel now uses looper's command line parser)
  - Fixed a bug that required re-starting with token upon any errors.
  - Allow for activating subprojects using new `peppy` method.
  
### Added:

  - Allow setting the authentication token in the caravel config file

## [0.1] -- 2018-12-21

### Added:

  - First semi-functional release of caravel.



caravel/const.py
""" Package constants """
import os
from divvy.const import COMPUTE_SETTINGS_VARNAME
from looper import __version__ as LOOPER_VERSION, Sample as LooperSample
from peppy import __version__ as PEPPY_VERSION
from ._version import __version__ as CARAVEL_VERSION
from ubiquerg import filesize_to_str
from warnings import warn


def get_req_version(module=None):
    """
    Read the required version of the specified module from the requirements file

    :param str module: the module to be checked
    :return dict | None: the required version and name of the requested module
    """
    reqs_file = os.path.join(os.path.dirname(__file__),
                             "requirements", "requirements-all.txt")

    if module is not None and os.path.isfile(reqs_file):
        with open(reqs_file) as rf:
            for l in rf:
                try:
                    p, v = l.split("=")
                except ValueError:
                    continue
                if module in p:
                    return {module: v.lstrip("=").rstrip("\n")}
            else:
                raise Exception("Requirement parse failed: {}".
                                format(reqs_file))
    else:
        raise IOError("The requirements file '{rf}' not found or module arg '{mod}' is missing. "
                      "The version of '{mod}' could not be asserted.".format(rf=reqs_file, mod=module))
        return None


def input_sizes(p):
    """
    Safely determine all the input file sizes for a project. If any exception raises, return "NA".

    :param looper.Project p: project to determine input file sizes for
    :return str: sum input file sizes
    """
    cumulative_file_size = 0
    try:
        for s in p.samples:
            s = LooperSample(s)
            for proto in [s.protocol] if isinstance(s.protocol, str) else s.protocol:
                for piface in p.get_interfaces(proto):
                    s.set_pipeline_attributes(piface, piface.fetch_pipelines(proto))
                    cumulative_file_size += s.input_file_size * 1024 ** 3
        return filesize_to_str(cumulative_file_size)
    except Exception as e:
        warn("Could not determine file size for project: '{}'. Got: '{}'".format(p.name, e.__class__.__name__))
        return "NA"


DEFAULT_PORT = 5000
DEFAULT_LOGGING_LVL = 20
CONFIG_ENV_VAR = "CARAVEL"
CONFIG_PRJ_KEY = "projects"
CONFIG_TOKEN_KEY = "token"
TOKEN_FILE_NAME = ".caravel_token"
EXAMPLE_FILENAME = "caravel_demo.yaml"
TOKEN_LEN = 15
SET_ELSEWHERE = [["--force-yes"], ["--sp"], ["--env"], ["--help"], ["--version"], ["--selector-attribute"],
                 ["--selector-exclude"], ["--selector-include"], ['--resources'], ['--compute-package'], ['']]
LOG_FILENAME = "caravel.log"
REQUIRED_LOOPER_VERSION = get_req_version("loopercli")["loopercli"]
REQUIRED_PEPPY_VERSION = get_req_version("peppy")["peppy"]
REQUIRED_V_BY_NAME = {"loopercli": REQUIRED_LOOPER_VERSION, "peppy": REQUIRED_PEPPY_VERSION}
V_BY_NAME = {"loopercli": LOOPER_VERSION, "peppy": PEPPY_VERSION, "caravel": CARAVEL_VERSION}
DEFAULT_TERMINAL_WIDTH = 80
SUMMARY_NAVBAR_PLACEHOLDER = "<li class='nav-item'><a class='nav-link disabled'>No summary yet</a></li>"
TEMPLATES_DIRNAME = "jinja_templates"
TEMPLATES_PATH = os.path.join(os.path.dirname(os.path.abspath(__file__)), TEMPLATES_DIRNAME)
DEMO_FILE_PATH = os.path.join(os.path.dirname(os.path.abspath(__file__)), EXAMPLE_FILENAME)
COMMAND_KEY = "execute"
POLL_INTERVAL = 3  # in seconds
MISSING_SAMPLE_DATA_TXT = "<code>looper run</code> was called, but not all the samples were correctly processed. " \
                            "</br>Possible reasons: <ul style='padding-left: 30px;'>"\
                            "<li>all jobs are still in a queue</li>" \
                            "<li>submission was not successful</li></ul>"
REQ_CFG_VERSION = 0.2
# this preferences/types can be set in the config file under "preferences" key
PREFERENCES_NAMES_TYPES = {"status_check_interval": int,
                           "compute_package": str}
# mapping of looper.Project metadata of interest and lambda expressions extracting them
PROJECT_MDATA_FUN = {"name": lambda p: p.name,
                     "names_sp": lambda p: ", ".join(p.subprojects.keys()),
                     "num_sp": lambda p: len(p.subprojects.keys()),
                     "num_samples": lambda p: p.num_samples,
                     "protocols": lambda p: ", ".join(sorted(p.protocols)),
                     "inputs_size": input_sizes}


"""
Config file structure determination 
"""
# config file structure related consts

CFG_NAME = "caravel configuration"
CFG_ENV_VARS = ["CARAVEL"]

CFG_VERSION_KEY = "config_version"
CFG_PROJECTS_KEY = "projects"
CFG_SUBPROJECTS_KEY = "subprojects"
CFG_PREFERENCES_KEY = "preferences"
CFG_PROJECT_NAME_KEY = "name"
CFG_PROJECT_DESC_KEY = "project_description"

CFG_EXAMPLE = """
# example {cfg_name} structure
{version}: 0.2
{preferences}:
  status_check_interval: 2

{projects}:
  /home/johndoe/projects/config_rnaseq.yaml:
    {name}: rnaseq_february
    {desc}: Processes a set of RNA-seq samples 
""".format(cfg_name=CFG_NAME, version=CFG_VERSION_KEY, projects=CFG_PROJECTS_KEY, preferences=CFG_PREFERENCES_KEY,
           name=CFG_PROJECT_NAME_KEY, desc=CFG_PROJECT_DESC_KEY)
